keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,589 @@ yet but we think may be interesting to consider in the future.
+too low for either.
+
+The configuration of an apiserver includes a concurrency limit.  This
+is a number, whose units is a number of readonly requests served
+concurrently.  Unlike in today's max-in-flight handler, the mutating
+and readonly requests are commingled without distinction.  The primary
+resource limit applied is that at any moment in time the number of
+running non-exempt requests should not exceed the concurrency limit.
+Requests of an exempt priority are neither counted nor limited, as in
+today's max-in-flight handler.  For the remainder, each server's
+overall concurrency limit is divided among those non-exempt priority
+levels and each enforces its own limit (independently of the other
+levels).
+
+At the first stage of development, an apiserver’s concurrency limit
+will be derived from the existing configuration options for
+max-mutating-in-flight and max-readonly-in-flight, by taking their
+sum.  Later we may migrate to a single direct configuration option.
+
+#### Secondary Memory Protection
+
+A RequestPriority is also configured with a limit on the number of
+requests that may be waiting in a given queue.
+
+#### Latency Protection
+
+An apiserver is also configured with a limit on the amount of time
+that a request may wait in its queue.  If this time passes while a
+request is still waiting for service then the request will be
+rejected.
+
+### Queuing
+
+Once a request is categorized and assigned to a queue the next
+decision is whether to reject or accept that request.
+
+A request of an exempt priority is never rejected and never waits in a
+queue; such a request is dispatched as soon as it arrives.
+
+For queuing requests of non-exempt priority, the first step is to
+reject all the requests that have been waiting longer than the
+configured limit.  Once that is done, the newly arrived request is
+considered.  This request is rejected if and only if the total number
+of requests waiting in its queue is at least the configured limit on
+that number.
+
+A possible alternative would accept the request unconditionally and,
+if that made the queue too long, reject the request at the head of the
+queue.  That would be the preferred design if we were confident that
+rejection will cause the client to slow down.  Lacking that
+confidence, we choose to reject the youngest rather than the oldest
+request of the queue, so that an investment in holding a request in a
+queue has a chance of eventually getting useful work done.
+
+### Dispatching
+
+Requests of an exempt priority are never held up in a queue; they are
+always dispatched immediately.  Following is how the other requests
+are dispatched at a given apiserver.
+
+The concurrency limit of an apiserver is divided among the non-exempt
+priority levels in proportion to their assured concurrency shares.
+This produces the assured concurrency value (ACV) for each non-exempt
+priority level:
+
+```
+ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
```

Comment:
My understanding of this equation with some #s:

SCL = 400
( sum[priority levels k] ACS(k) ) = 800 (eight priority levels, assuredConcurrencyShares:100 for all)
ACS(l) = 100 (priority level l, assuredConcurrencyShares:100)
ACV(l) = ceil(400 * 100/800)
ACV(l) = 50

Is this understanding correct? That as ( sum[priority levels k] ACS(k) ) increases the ACV for all priority levels decreases?  Also is a priority level each individual RequestPriority?

Reply:
I agree with the values you calculated in that example.  As for your general observation, note that `( sum[priority levels k] ACS(k) )` can increase only if some individual `ACS(k)` values increase.  If they all increase in the same proportion, then there is no change in the result.  If they change in different proportions, then the distribution shifts around.

Yes, modulo defaulting, there is a one-to-one correspondence between the intuitive notion of request level and RequestPriority API objects.